[AMORO-3272] data-expire by partition info

lintingbin opened the pull request:
<!--
Thanks for sending a pull request!

Here are some tips for you:
  1. If this is your first time, please read our contributor guidelines: https://amoro.apache.org/how-to-contribute/
  2. If the PR is related to an issue in https://github.com/apache/amoro/issues, add '[AMORO-XXXX]' in your PR title, e.g., '[AMORO-XXXX] Your PR title ...'.
  3. If the PR is unfinished, add '[WIP]' in your PR title, e.g., '[WIP][AMORO-XXXX] Your PR title ...'.
-->

## Why are the changes needed?
<!--
Please clarify why the changes are needed. For instance,
  1. If you add a feature, you can talk about its use case.
  2. If you fix a bug, you can clarify why it is a bug.
  3. Use Fix/Resolve/Close #{ISSUE_NUMBER} to link this PR to its related issue
-->

Close #3272.

## Brief change log
<!--
Clearly describe the changes made in modules, classes, methods, etc.
-->

- When the expiration field is the partition field and the expiration level is partition, prioritize using the partition information of the datafile to expire the data files.

- Modify the expected results of a test case, which is slightly different from the previous implementation. In partition expiration, date comparison is needed, and data files with the same date should not be expired.

## How was this patch tested?

- [x] Add some test cases that check the changes thoroughly including negative and positive cases if possible

- [ ] Add screenshots for manual tests if appropriate

- [x] Run test locally before making a pull request

## Documentation

- Does this pull request introduce a new feature? no
- If yes, how is the feature documented? (not applicable / docs / JavaDocs / not documented)

klion26 commented on amoro-ams/src/main/java/org/apache/amoro/server/optimizing/maintainer/IcebergTableMaintainer.java:
```diff
@@ -696,26 +702,152 @@ CloseableIterable<FileEntry> fileScan(
+
+  private Comparable<?> getPartitionUpperBound(
+      DataExpirationConfig expirationConfig, Types.NestedField field, long expireTimestamp) {
+    switch (field.type().typeId()) {
```
could you please add some comments on this?  especially for the type `TIMESAMPT` and `LONG`

klion26 commented on amoro-ams/src/main/java/org/apache/amoro/server/optimizing/maintainer/IcebergTableMaintainer.java:
```diff
@@ -696,26 +702,152 @@ CloseableIterable<FileEntry> fileScan(
+        .forEach(expiredFiles::addFile);
+  }
+
+  private Map<Integer, Map<Integer, PartitionField>> buildExpirePartitionFieldsMap(
```
Do we need to add some comments to this function here, such as what is the key/value of the return value, and the brief logic

klion26 commented on amoro-ams/src/main/java/org/apache/amoro/server/optimizing/maintainer/IcebergTableMaintainer.java:
```diff
@@ -696,26 +702,152 @@ CloseableIterable<FileEntry> fileScan(
+              boolean expired = partitionValue.compareTo(entry.getValue()) < 0;
+              expiredList.add(expired);
+            }
+            if (!expiredList.isEmpty() && expiredList.stream().allMatch(Boolean::booleanValue)) {
```
why do we need to add the `expiredList.stream().allMatch` check here

lintingbin replied:
It's possible that a timestamp field with event_time is configured with multiple partition fields, such as year(event_time), month(event_time), and day(event_time). All these conditions need to be met together to delete a partition.

klion26 commented on amoro-ams/src/test/java/org/apache/amoro/server/optimizing/maintainer/TestDataExpire.java:
```diff
@@ -194,17 +194,11 @@ private void testUnKeyedPartitionLevel() {
-                createRecord(3, "333", parseMillis("2022-01-02T12:00:00"), "2022-01-02T12:00:00"),
-                createRecord(4, "444", parseMillis("2022-01-02T19:00:00"), "2022-01-02T19:00:00"));
-      }
+      expected =
```
It seems the table was partitioned by `op_time` with `day` transformer, and will expire the data before 86400s, so the record created in `2022-01-02` and `2022-01-03` will still be there(because the partition can't be expired) is this understanding correct?

In the other way, is there any way we can know the partition for each record easier?

klion26 replied:
can we add some comments to describe them here? thanks

klion26 commented on amoro-ams/src/test/java/org/apache/amoro/server/optimizing/maintainer/TestDataExpire.java:
```diff
@@ -194,17 +194,11 @@ private void testUnKeyedPartitionLevel() {
+              createRecord(4, "444", parseMillis("2022-01-02T19:00:00"), "2022-01-02T19:00:00"));
     } else {
       expected =
           Lists.newArrayList(
```
It seems that if the table is not partitioned, we would not expire any records. do I understand right?

lintingbin replied:
Yes, under this PartitionLevel test case.

klion26 reviewed (commented):
@lintingbin Thanks for the contribution, I left some comments, please take a look when you're free thanks.

lintingbin commented on amoro-ams/src/main/java/org/apache/amoro/server/optimizing/maintainer/IcebergTableMaintainer.java:
```diff
@@ -696,26 +702,152 @@ CloseableIterable<FileEntry> fileScan(
+        expireByPartitionSuccess =
+            tryExpireByPartition(entries, expirationConfig, expireTimestamp, expiredFiles);
+      }
+      if (!expireByPartitionSuccess) {
```
Yes

lintingbin commented on amoro-ams/src/main/java/org/apache/amoro/server/optimizing/maintainer/IcebergTableMaintainer.java:
```diff
@@ -696,26 +702,152 @@ CloseableIterable<FileEntry> fileScan(
+        buildExpirePartitionFieldsMap(expirationField);
+    // All historical specs have expirationField as the partition field.
+    boolean allSpecsMatch = expirePartitionFieldsMap.values().stream().noneMatch(Map::isEmpty);
+    if (allSpecsMatch) {
```
If the partition field has been modified and the data-expire configuration is set to the modified partition field, previous data might not be deleted. Therefore, we have implemented a strict check that requires all historical partition fields to meet the conditions; otherwise, it will revert to metric deletion.

lintingbin commented:
@klion26 I have already responded and made modifications. Please help review it again.

klion26 commented:
@lintingbin thanks for rebasing the comments, the change LGTM. let's see if there are any more comments from the community.

Do we need to modify the corresponding documents?

lintingbin commented:
> @lintingbin thanks for rebasing the comments, the change LGTM. let's see if there are any more comments from the community.
> 
> Do we need to modify the corresponding documents?

There is no need to modify the documentation since there are no changes in the parameters.

XBaith commented on amoro-ams/src/main/java/org/apache/amoro/server/optimizing/maintainer/IcebergTableMaintainer.java:
```diff
@@ -696,26 +702,155 @@ CloseableIterable<FileEntry> fileScan(
+            List<Boolean> expiredList = new ArrayList<>();
+            ContentFile<?> contentFile = fileEntry.getFile();
+            int fileSpecId = contentFile.specId();
+            for (Map.Entry<Integer, Comparable<?>> entry :
```
Could you rename this variable to something more readable? like `posToValue`

XBaith commented on amoro-ams/src/main/java/org/apache/amoro/server/optimizing/maintainer/IcebergTableMaintainer.java:
```diff
@@ -696,26 +702,155 @@ CloseableIterable<FileEntry> fileScan(
+            posToField.getKey(),
+            ((SerializableFunction<Comparable<?>, Comparable<?>>)
+                    posToField.getValue().transform().bind(field.type()))
+                .apply(expireValue));
```
What will happen if the partition value type is `Void`, it is comparable?

XBaith commented on amoro-ams/src/main/java/org/apache/amoro/server/optimizing/maintainer/IcebergTableMaintainer.java:
```diff
@@ -696,26 +702,155 @@ CloseableIterable<FileEntry> fileScan(
+      Types.NestedField field,
+      Comparable<?> expireValue) {
+    Map<Integer, Map<Integer, Comparable<?>>> expirePartitionValue = new HashMap<>();
+    for (Map.Entry<Integer, Map<Integer, PartitionField>> entry :
```
Need to be more readable

lintingbin commented:
@XBaith I have refactored the code by removing some unnecessary parts and optimizing the naming of variables to enhance readability. I also added a check for cases where transform is Void. Please review the code again.

XBaith commented:
> @XBaith I have refactored the code by removing some unnecessary parts and optimizing the naming of variables to enhance readability. I also added a check for cases where transform is Void. Please review the code again.

I don't see any unit test for that case. Do you mean you test locally for `Void` transform?

XBaith reviewed (changes requested):
I wrote some unit tests for a case involving expiring partitions after dropping the partition field. Unfortunately, the current procedure cannot handle this scenario correctly.

Here’s the partition level case:

- I set the partition field as `op_time` and inserted some records.
- Then, I removed the partition field before expiring the data.

**Expected behavior**: All records should be retained since the partition field has been removed.

Please fix this issue and add additional unit tests to cover similar scenarios. Thanks!

XBaith commented:
Sorry for the mistake; this bug existed prior to this PR. I will raise another PR to address it.  

Additionally, we should add more unit tests to cover vulnerabilities as much as possible. Thanks for your contribution!

klion26 reviewed (approved):
Thanks for the contribution, LGTM

XBaith reviewed (approved):
LGTM

czy006 commented on amoro-ams/src/test/java/org/apache/amoro/server/optimizing/maintainer/TestDataExpire.java:
```diff
@@ -591,6 +592,7 @@ protected static Map<String, String> getDefaultProp() {
     prop.put(TableProperties.ENABLE_DATA_EXPIRATION, "true");
     prop.put(TableProperties.DATA_EXPIRATION_FIELD, "op_time");
     prop.put(TableProperties.DATA_EXPIRATION_RETENTION_TIME, "1d");
+    prop.put("write.metadata.metrics.default", "none");
```
Does "write.metadata.metrics.default" need to be added to TableProperties ? @XBaith

XBaith replied:
You are right. It should not be the default property. It's a good idea to create a new test case to check that `write.metadata.metrics.default` is `none` to trigger data expiration based on partition informatio

lintingbin replied:
You're right, I should create a set of test cases with "write.metadata.metrics.default" set to "none." However, I'm currently stuck with the issue that configuring "write.metadata.metrics.default" as "none" in the test cases doesn't take effect. Even when I set "write.metadata.metrics.default" to "none," the upper_bound is still being recorded in the metrics. Could you help me take a look?

lintingbin replied:
During debugging, I found that the relevant configuration for metrics was not correctly passed when writing Parquet files, leading to "none" not taking effect. However, in my production environment, there is no issue.

<img width="1432" alt="image" src="https://github.com/user-attachments/assets/b935ad1b-4640-48ce-9995-ba28576757de" />
<img width="1134" alt="image" src="https://github.com/user-attachments/assets/b82a8843-cb15-4a51-9277-fe4ae1b1c542" />

XBaith replied:
Looks like iceberg GenericAppenderFactory do not put the right metric configs. Currently, I think we can just add a new test for setting metric property as `none`, then I will report the issue and try to fix it in Iceberg.

lintingbin commented on amoro-ams/src/main/java/org/apache/amoro/server/optimizing/maintainer/IcebergTableMaintainer.java:
```diff
@@ -917,17 +955,23 @@ static boolean willNotRetain(
     Object upperBound =
         Conversions.fromByteBuffer(type, contentFile.upperBounds().get(field.fieldId()));
     Literal<Long> literal = Literal.of(Long.MAX_VALUE);
+    if ("none".equals(table.properties().get("write.metadata.metrics.default"))) {
+      upperBound = null;
+    }
```
@XBaith @czy006 I have added test cases, but without these three lines of code, two test cases fail. However, these three lines are currently unnecessary and are only used to solve the issue of the write.metadata.metrics.default configuration not being effective. I now need to have the test cases pass while removing these three lines of code.

lintingbin commented:
@czy006 @XBaith Now that all the problems have been solved, please help to review the code again.

XBaith commented:
I made minor code adjustments to ensure that the expiration behavior remains consistent when the expire field is of string type, preventing any confusion for users
